lib/HLSL/DxilPreparePasses.cpp

```diff
@@ -376,6 +376,80 @@ class DxilFinalizeModule : public ModulePass {
         CI->eraseFromParent();
       }
     }
+
+    // Replace llvm.lifetime.start/.end intrinsics with undef stores.
+    // This works around losing scoping information in earlier shader models
+    // that do not support the intrinsics natively.
+
+    // Get the declarations. This may introduce them if there were none before.
+    Value *StartDecl = Intrinsic::getDeclaration(&M, Intrinsic::lifetime_start);
+    Value *EndDecl   = Intrinsic::getDeclaration(&M, Intrinsic::lifetime_end);
+
+    // Collect all calls to both intrinsics.
+    std::vector<CallInst*> intrinsicCalls;
+    for (Use &U : StartDecl->uses()) {
+      // All users must be call instructions.
+      CallInst *CI = dyn_cast<CallInst>(U.getUser());
+      DXASSERT(CI, "Expected user of lifetime.start intrinsic to be a CallInst");
+      intrinsicCalls.push_back(CI);
+    }
+    for (Use &U : EndDecl->uses()) {
+      // All users must be call instructions.
+      CallInst *CI = dyn_cast<CallInst>(U.getUser());
+      DXASSERT(CI, "Expected user of lifetime.end intrinsic to be a CallInst");
+      intrinsicCalls.push_back(CI);
+    }
+
+    // Replace each intrinsic with an undef store.
+    for (CallInst *CI : intrinsicCalls) {
+      // Find the corresponding pointer (can be alloca, global value, an
+      // argument, ...).
+      Value *voidPtr = CI->getArgOperand(1);
+      DXASSERT(voidPtr->getType()->isPointerTy() &&
+               voidPtr->getType()->getPointerElementType()->isIntegerTy(8),
+               "Expected operand of lifetime intrinsic to be of type i8*" );
+
+      Value *ptr = nullptr;
+      if (BitCastInst *BC = dyn_cast<BitCastInst>(voidPtr)) {
+        ptr = BC->getOperand(0);
+      } else if (ConstantExpr *CE = dyn_cast<ConstantExpr>(voidPtr)) {
```

Comment:
It does. This can happen if a local variable/array is promoted to a constant global. I think lifetime intrinsics are perfectly fine for constant globals even though they may not be of much value. An example where this happens is the float3 arrays DirLightDirections and DirLightColors in HLSLFileCheck/samples/d3d11/SubD11_SmoothPS.hlsl or in the various hoisting tests in HLSLFileCheck/hlsl/types/modifiers/const.